### DISCUSSION TITLE
Anyway to use Drift with Deposit package?

### DISCUSSION THREAD

**Discussion by adar2378** [simolus3/drift#2591] [Q&A]
I want to make my app backend agnostic. It should work with remote or local backends. Can I make it work using Drift in DRY way? I want to create a single [DepositAdapter](https://github.com/bluefireteam/deposit/blob/main/packages/deposit/lib/src/deposit_adapter.dart) that will work for any tables from drift? So far I can understand that I can do it via separete DepositAdapter based on separate tables via those table related DAOs. Any DRY solution or idea would be very helpful. Thanks!

https://github.com/bluefireteam/deposit/tree/main/packages/deposit

**Comment by simolus3** [comment]
Drift is pretty fundamentally tied to relational databases. At the moment the main focus is sqlite3, but many of the features supported by drift also work with PostgreSQL or MariaDB.

I haven't looked at deposit in-depth, but it looks like it's abstracting over document stores instead of relational databases. These are solving slightly different problems and they're not able to answer some queries you might formulate with drift (in particular, things like aggregates or groups are difficult to implement efficiently).
So it seems unlikely that you can use deposit as a backend for drift - but the other way around will likely work pretty well (using deposit with drift or sqlite3 as a backend).

**Reply by adar2378** [reply]
Yeah I want to use Drift as one of the backends. I plan to add remote backends as well. So far I found something like this as a solution. I was wondering if it can be implemented without using the tables mapper. Thank you for your previous answer.

```dart

class DriftAdapter implements DepositAdapter<int> {
  final AppDatabase appDatabase;

  DriftAdapter({required this.appDatabase});

  final Map<String, _TableData> _tables = {
    'projects': _TableData(
      table: appDatabase.projects,
      converter: (data) => ProjectsCompanion.fromData(data),
    ),
    'tasks': _TableData(
      table: appDatabase.tasks,
      converter: (data) => TasksCompanion.fromData(data),
    ),
    // ... add other tables and their converters here
  };

  _TableData _getTableData(String tableName) {
    final data = _tables[tableName];
    if (data == null) {
      throw Exception('Table $tableName not found');
    }
    return data;
  }

  @override
  Future<Map<String, dynamic>> add(
      String table, String primaryColumn, Map<String, dynamic> data) async {
    final tableData = _getTableData(table);
    final id = await appDatabase.into(tableData.table).insert(tableData.converter(data));
    return {...data, primaryColumn: id};
  }

  // ... Implement other methods similarly

  // ... Continue with other methods
}

class _TableData {
  final TableInfo table;
  final Insertable Function(Map<String, dynamic> data) converter;

  _TableData({required this.table, required this.converter});
}

```

**Reply by simolus3** [reply]
Do you only want to use drift as a deposit backend or do you have an existing drift database you want to make available with deposit? In the first case, you can define a drift database that matches the general schema (using a JSON structure and a text column for data) and use that table all the time (you can call `alias` to refer to the same table under different names, even for inserts). If you have existing tables and you want to keep using their schema, I think you need to include the table references.
